integrations/linear/src/schemas.ts

```diff
@@ -0,0 +1,423 @@
+  type: z.literal("Issue"),
+  data: IssueDataSchema,
+});
+export const IssueEventSchema = z.discriminatedUnion("action", [
+  IssueEventBaseSchema.extend({
+    action: CREATE,
+  }),
+  IssueEventBaseSchema.extend({
+    action: REMOVE,
+  }),
+  IssueEventBaseSchema.extend({
+    action: UPDATE,
+    updatedFrom: IssueDataSchema.partial(),
+  }),
+]);
+export type IssueEvent = z.infer<typeof IssueEventSchema>;
+
+export const IssueLabelEventBaseSchema = WebhookPayloadBaseSchema.extend({
+  type: z.literal("IssueLabel"),
+  data: IssueLabelDataSchema,
+});
+export const IssueLabelEventSchema = z.discriminatedUnion("action", [
+  IssueLabelEventBaseSchema.extend({
+    action: CREATE,
+  }),
+  IssueLabelEventBaseSchema.extend({
+    action: REMOVE,
+  }),
+  IssueLabelEventBaseSchema.extend({
+    action: UPDATE,
+    updatedFrom: IssueLabelDataSchema.partial(),
+  }),
+]);
+export type IssueLabelEvent = z.infer<typeof IssueLabelEventSchema>;
+
+// TODO: confirm this with real-world payload(s)
+export const IssueSLAEventSchema = WebhookPayloadBaseSchema.extend({
+  type: z.literal("IssueSLA"),
+  action: z.union([z.literal("set"), z.literal("highRisk"), z.literal("breached")]),
+  issueData: IssueDataSchema,
+});
+export type IssueSLAEvent = z.infer<typeof IssueSLAEventSchema>;
+
+export const ProjectEventBaseSchema = WebhookPayloadBaseSchema.extend({
+  type: z.literal("Project"),
+  data: ProjectDataSchema,
+});
+export const ProjectEventSchema = z.discriminatedUnion("action", [
+  ProjectEventBaseSchema.extend({
+    action: CREATE,
+  }),
+  ProjectEventBaseSchema.extend({
+    action: REMOVE,
+  }),
+  ProjectEventBaseSchema.extend({
+    action: UPDATE,
+    updatedFrom: ProjectDataSchema.partial(),
+  }),
+]);
+export type ProjectEvent = z.infer<typeof ProjectEventSchema>;
+
+export const ProjectUpdateEventBaseSchema = WebhookPayloadBaseSchema.extend({
+  type: z.literal("ProjectUpdate"),
+  data: ProjectUpdateDataSchema,
+});
+export const ProjectUpdateEventSchema = z.discriminatedUnion("action", [
+  ProjectUpdateEventBaseSchema.extend({
+    action: CREATE,
+  }),
+  ProjectUpdateEventBaseSchema.extend({
+    action: REMOVE,
+  }),
+  ProjectUpdateEventBaseSchema.extend({
+    action: UPDATE,
+    updatedFrom: ProjectUpdateDataSchema.partial(),
+  }),
+]);
+export type ProjectUpdateEvent = z.infer<typeof ProjectUpdateEventSchema>;
+
+export const ReactionEventBaseSchema = WebhookPayloadBaseSchema.extend({
+  type: z.literal("Reaction"),
+  data: ReactionDataSchema,
+});
+export const ReactionEventSchema = z.discriminatedUnion("action", [
+  ReactionEventBaseSchema.extend({
+    action: CREATE,
+  }),
+  ReactionEventBaseSchema.extend({
+    action: REMOVE,
+  }),
+  ReactionEventBaseSchema.extend({
+    action: UPDATE,
+    updatedFrom: ReactionDataSchema.partial(),
+  }),
+]);
+export type ReactionEvent = z.infer<typeof ReactionEventSchema>;
+
+export const WebhookPayloadSchema = z.union([
```

Comment:
This was once a discriminated union (as it should be), but there were issues with nesting. See:
- https://github.com/colinhacks/zod/issues/1618
- https://github.com/colinhacks/zod/issues/2523

Even with the proposed workarounds it always ended up breaking somewhere.